modules/gmake/gmake.lua

```diff
@@ -0,0 +1,338 @@
+
+	table.sort(prj.configs, function(cfg0, cfg1)
+		return (cfg0.gmake_identifier:lower() < cfg1.gmake_identifier:lower())
+	end)
+
+	return prj
+end
+
+
+---
+-- Fetch the settings for a specific workspace-level configuration.
+--
+-- @param wks
+--	The `dom.Workspace` instance which contains the target configuration.
+-- @param build
+--	The target build configuration name, eg. 'Debug'
+-- @param platform
+--	The target platform name.
+-- @returns
+--	A `dom.Config`, with additional Visual Studio specific values.
+---
+function gmake.fetchWorkspaceConfig(wks, build, platform)
+	local cfg = gmake.fetchConfig(wks
+		:selectAny({ configurations = build, platforms = platform })
+		:fromScopes(wks.root)
+		:withInheritance()
+	)
+
+	cfg.root = wks.root
+	cfg.workspace = wks
+
+	return cfg
+end
+
+
+---
+-- Fetch the settings for a specific project-level configuration.
+--
+-- @param wks
+--	The `dom.Project` instance which contains the target configuration.
+-- @param build
+--	The target build configuration name, eg. 'Debug'
+-- @param platform
+--	The target platform name.
+-- @returns
+--	A `dom.Config`, with additional Visual Studio specific values.
+---
+function gmake.fetchProjectConfig(prj, build, platform)
+	local cfg = gmake.fetchConfig(prj
+		:selectAny({ configurations = build, platforms = platform })
+		:fromScopes(prj.root, prj.workspace)
+	)
+
+	cfg.root = prj.root
+	cfg.workspace = prj.workspace
+	cfg.project = prj
+
+	-- TODO: Compiler/Linker from DOM
+	cfg.compiler = gmake.toolsets.compilers.gcc
+	cfg.linker = gmake.toolsets.linkers.gcc
+
+	-- Configurations inherit most values from project, but files should be kept separated
+	cfg.files = cfg:withoutInheritance().files
+
+	return cfg
+end
+
+
+---
+-- Fetch the settings for a specific file-level configuration.
+--
+-- @param cfg
+--	The `dom.Config` instance which contains the target file settings. May be a
+--	workspace or project configuration.
+-- @param file
+--	The path of the file for which settings should be fetched.
+-- @returns
+--	A `dom.Config`, with additional Visual Studio specific values.
+---
+function gmake.fetchFileConfig(cfg, file)
+	local fileCfg = dom.Config.new(cfg
+		:select({ files = file })
+		:fromScopes(cfg.root, cfg.workspace, cfg.project)
+	)
+
+	fileCfg.file = file
+	fileCfg.project = cfg.project
+	fileCfg.gmake_build = cfg.gmake_build
+
+	return fileCfg
+end
+
+
+---
+-- Helper for `gmake.fetchWorkspaceConfig()` and `gmake.fetchProjectConfig()`.
+-- Computes common Visual Studio specific values required by the exporter.
+---
+function gmake.fetchConfig(state)
+	local cfg = dom.Config.new(state)
+
+	-- translate the incoming architecture
+	cfg.gmake_architecture = _ARCHITECTURES[cfg.architecture] or 'x86_64'
+	cfg.platform = cfg.platform or cfg.gmake_architecture
+
+	-- "Configuration|Platform or Architecture", e.g. "Debug|MyPlatform" or "Debug|Win32"
+	cfg.gmake_identifier = string.format('%s|%s', cfg.configuration, cfg.platform)
```

Comment:
Probably don't want to use `|` here? That's a Visual Studio convention. FWIW I'm using space for the Xcode exporter; maybe `_` for makefiles so we don't have to escape the resulting strings when exporting? Also if you lowercase it here you won't have to do it when exporting it.